// zombie/lib/zombie/eventloop.js
// Generated by CoffeeScript 1.3.1
var EventLoop, URL, ms, raise;

ms = require("ms");

URL = require("url");

raise = require("./scripts").raise;

EventLoop = (function() {

  EventLoop.name = 'EventLoop';

  function EventLoop(_browser) {
    this._browser = _browser;
    this._processing = 0;
    this._waiting = [];
    this._timers = [];
  }

  EventLoop.prototype.reset = function() {
    var timer, _i, _len, _ref;
    if (this._timers) {
      _ref = this._timers;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        timer = _ref[_i];
        global.clearTimeout(timer.handle);
      }
    }
    return this._timers = [];
  };

  EventLoop.prototype.apply = function(window) {
    var execute,
      _this = this;
    execute = function(scope, code, notice) {
      _this._browser.log(notice);
      window._evaluate(code);
      return _this._next();
    };
    window.setTimeout = function(fn, delay) {
      var timer;
      delay = Math.max(delay || 0, 1);
      timer = {
        start: function() {
          timer.next = Date.now() + delay;
          if (_this._waiting.length > 0) {
            return timer.resume();
          }
        },
        resume: function() {
          var _this = this;
          if (timer.next) {
            timer.next = Date.now() + delay;
            return timer.handle = global.setTimeout(function() {
              delete timer.next;
              return execute("Timeout", fn, "Firing timeout after " + delay + "ms delay");
            }, timer.next - Date.now());
          }
        },
        pause: function() {
          delay = timer.next - Date.now();
          return global.clearTimeout(timer.handle);
        },
        stop: function() {
          timer.pause();
          return delete timer.next;
        }
      };
      timer.start();
      _this._timers.push(timer);
      return timer;
    };
    window.setInterval = function(fn, interval) {
      var timer;
      timer = {
        start: function() {
          timer.next = true;
          if (_this._waiting.length > 0) {
            return timer.resume();
          }
        },
        resume: function() {
          var _this = this;
          if (timer.next) {
            timer.handle = global.setInterval(function() {
              timer.next = Date.now() + interval;
              return execute("Interval", fn, "Firing interval every " + interval + "ms");
            }, interval);
            return timer.next = Date.now() + interval;
          }
        },
        pause: function() {
          return global.clearInterval(timer.handle);
        },
        stop: function() {
          timer.pause();
          return delete timer.next;
        }
      };
      timer.start();
      _this._timers.push(timer);
      return timer;
    };
    window.clearTimeout = function(timer) {
      if (timer && timer.stop && timer.next) {
        return timer.stop();
      }
    };
    return window.clearInterval = function(timer) {
      if (timer && timer.stop && timer.next) {
        return timer.stop();
      }
    };
  };

  EventLoop.prototype.perform = function(fn) {
    var _this = this;
    ++this._processing;
    fn(function() {
      --_this._processing;
      if (_this._processing === 0) {
        return _this._next();
      }
    });
  };

  EventLoop.prototype.dispatch = function(target, event) {
    var preventDefault;
    preventDefault = false;
    this.perform(function(done) {
      var window;
      window = (target.ownerDocument || target.document).window;
      window._evaluate(function() {
        return preventDefault = target.dispatchEvent(event);
      });
      return done();
    });
    return preventDefault;
  };

  EventLoop.prototype.wait = function(window, duration, callback) {
    var done, done_at, is_done, terminate, waiting,
      _this = this;
    if (typeof duration === "function") {
      is_done = duration;
      done_at = Infinity;
    } else {
      if (!(duration && duration !== 0)) {
        duration = this._browser.waitFor;
      }
      done_at = Date.now() + ms(duration || 0);
    }
    done = function(error) {
      var fn;
      done = null;
      _this._waiting = (function() {
        var _i, _len, _ref, _results;
        _ref = this._waiting;
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          fn = _ref[_i];
          if (fn !== waiting) {
            _results.push(fn);
          }
        }
        return _results;
      }).call(_this);
      if (_this._waiting.length === 0) {
        _this._pause();
      }
      if (terminate) {
        clearTimeout(terminate);
      }
      if (callback) {
        process.nextTick(function() {
          return callback(error, window);
        });
      }
      if (error) {
        return _this._browser.emit("error", error);
      } else {
        return _this._browser.emit("done");
      }
    };
    terminate = setTimeout(done, ms(this._browser.maxWait));
    waiting = function() {
      var next, timer, timers;
      if (!done) {
        return;
      }
      if (_this._processing > 0) {
        return;
      }
      try {
        if (is_done && is_done(window)) {
          done();
          return;
        }
      } catch (error) {
        done(error);
        return;
      }
      timers = (function() {
        var _i, _len, _ref, _results;
        _ref = this._timers;
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          timer = _ref[_i];
          if (timer.next) {
            _results.push(timer.next);
          }
        }
        return _results;
      }).call(_this);
      next = Math.min.apply(Math, timers);
      if (next > done_at) {
        return done();
      }
    };
    if (this._waiting.length === 0) {
      this._resume();
    }
    this._waiting.push(waiting);
    this._next();
  };

  EventLoop.prototype._next = function() {
    var waiting, _i, _len, _ref, _results;
    _ref = this._waiting;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      waiting = _ref[_i];
      _results.push(process.nextTick(waiting));
    }
    return _results;
  };

  EventLoop.prototype._pause = function() {
    var timer, _i, _len, _ref, _results;
    _ref = this._timers;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      timer = _ref[_i];
      if (timer.next) {
        _results.push(timer.pause());
      }
    }
    return _results;
  };

  EventLoop.prototype._resume = function() {
    var timer, _i, _len, _ref, _results;
    _ref = this._timers;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      timer = _ref[_i];
      if (timer.next) {
        _results.push(timer.resume());
      }
    }
    return _results;
  };

  EventLoop.prototype.dump = function() {
    return ["The time:   " + (new Date), "Timers:     " + this._timers.length, "Processing: " + this._processing, "Waiting:    " + this._waiting.length];
  };

  return EventLoop;

})();

module.exports = EventLoop;
